### ISSUE TITLE
free_deleter_impl should declare copy/assign as deleted

### ISSUE THREAD

**Issue by thongpv87** [scylladb/seastar#464]
https://github.com/scylladb/seastar/blob/ded9502c13f408abbdd96bd77758d5f68f8486b0/core/deleter.hh#L187-L191

The `operator=` is not deleted in this code. It can lead to double freeing of freed memory.

**Comment by denesb** [comment]
Having a user defined destructor will cause none of the copy/move constructors assignment operators to be generated.

**Comment by duarten** [comment]
Having a user-defined destructor does not prevent the copy constructor and copy assignment operator from being defaulted, so I don't think we should close this.

The upper layer (which uses the impl) does delete the copy special members, so this is not urgent, but it would indeed be more correct to delete them in `free_deleter_impl` too.

**Comment by denesb** [comment]
On 2018-06-19 14:13, Duarte Nunes wrote:
> Having a user-defined destructor does not prevent the copy constructor and copy assignment operator from being defaulted, so I don't think we should close this.
> 

You mean it doesn't prevent a programmer from adding a defaulted copy 
constructor and copy assignment operator? Nor does a deleted one. That 
said I'm not against adding deleting these to make the intent explicit.

> The upper layer (which uses the impl) does delete the copy special members, so this is not urgent, but it would indeed be more correct to delete them in `free_deleter_impl` too.
> 
How can external code delete this class's copy constructors and copy 
assignment operator? Or do they inherit from it?

**Comment by denesb** [comment]
On 2018-06-19 14:23, Duarte Nunes wrote:
> I mean that having a destructor does not prevent copy special members from being generated (as they were written with `= default`).


Yes but the above linked `free_deleter_impl` doesn't have any of those 
defaulted, hence they won't be generated.

> 
> I don't understand you last questions, but I'll try to clarify what I meant. Currently, we have this:
> 
> ```
> class deleter final {
>    impl* _impl = nullptr;
> public:
>    deleter() = default;
>     ~deleter();
>    deleter(const deleter&) = delete;
>    deleter& operator=(deleter&) = delete;
> };
> ```
> 
> Users never see `impl`, so they cannot copy it. Because `deleter` (the upper layer) deletes its copy special members, `impl` is also not copied internally. Thus, it is not strictly required to delete it's copy special members, because they are not invoked.
> 

I see.

**Comment by duarten** [comment]
Yes, they will be generated. Having a user-defined dtor doesn't prevent copy special members form being generated.

**Comment by denesb** [comment]
On 2018-06-19 14:34, Duarte Nunes wrote:
> Yes, they will be generated. Having a user-defined dtor doesn't prevent copy special members form being generated.
> 

_goes away to fiddle with compiler explorer_

It seems that you are right. I don't know why I though a user defined 
destructor will cause these not to be generated.

@thongpv3 feel free to send a patch.

**Comment by duarten** [comment]
You should consider tattooing https://i.stack.imgur.com/b2VBV.png, but avoid the mistake I made of having it done on my back.

**Comment by denesb** [comment]
On 2018-06-19 14:43, Duarte Nunes wrote:
> You should consider tattooing https://i.stack.imgur.com/b2VBV.png, but avoid the mistake I made of having it done on my back.
> 

You gotta love C++ its such a nice and simple language. :)))
I'll print this and hang it on my wall.

**Comment by xemul** [comment]
e332d3cc540b691278d0142445215950b9ac4b47
